File: examples/2015-day-01/go/aoc/aoc.go
package aoc

import "fmt"

func SolvePartOne(input string) int {
	floor := 0

	for _, char := range input {
		floor += step(char)
	}

	return floor
}

func SolvePartTwo(input string) int {
	floor := 0

	for i, char := range input {
		floor += step(char)

		if floor == -1 {
			return 1 + i
		}
	}

	return 0
}

func step(char rune) int {
	switch char {
	case '(':
		return 1
	case ')':
		return -1
	default:
		panic(fmt.Sprintf("unexpected character %q", char))
	}
}

File: examples/2015-day-01/swift/Sources/AoC/AoC.swift
public func solvePartOne(_ input: String) -> Int {
  var floor = 0

  for char in input {
    switch char {
    case "(":
      floor += 1
    case ")":
      floor -= 1
    default:
      fatalError("Invalid char \(char)")
    }
  }

  return floor
}

public func solvePartTwo(_ input: String) -> Int {
  var floor = 0

  for (i, char) in input.enumerated() {
    switch char {
    case "(":
      floor += 1
    case ")":
      floor -= 1
    default:
      fatalError("Invalid char \(char)")
    }

    if floor == -1 {
      return 1 + i
    }
  }

  return 0
}
